File: src/daemon/src/drivers/vendors/corsair/protocols/corsair_nxp.rs
// SPDX-License-Identifier: GPL-2.0-or-later
// SPDX-FileCopyrightText: Adam Honse (CalcProgrammer1) — OpenRGB project

//! Corsair "NXP" peripheral wire protocol.
//!
//! Device-agnostic: the caller supplies the device class byte and the color
//! buffer. See `docs/protocols/corsair-nxp.md`.

use anyhow::Result;

use crate::drivers::transports::{hid::HidTransport, Transport};
use halod_shared::types::RgbColor;

/// Report buffer: byte 0 is the `0x00` report ID, bytes 1..65 the 64-byte payload.
pub const REPORT_SIZE: usize = 65;
const TIMEOUT_MS: i32 = 1000;

const CMD_WRITE: u8 = 0x07;
const CMD_READ: u8 = 0x0E;
const CMD_STREAM: u8 = 0x7F;

const PROP_FIRMWARE: u8 = 0x01;
const PROP_SPECIAL_FUNCTION: u8 = 0x04;
const PROP_LIGHTING_CONTROL: u8 = 0x05;
const PROP_LAYOUT_SETUP: u8 = 0x40;
const PROP_SUBMIT_KEYBOARD_24: u8 = 0x28;

const LAYOUT_SETUP_SUB: u8 = 0x1E;
const LAYOUT_PRIMER_MODE: u8 = 0x08;
const LAYOUT_KEY_TRAILER: u8 = 0xC0;
const LAYOUT_KEYS_PER_PACKET: usize = 30;
const LAYOUT_PACKET_COUNT: usize = 4;

const HARDWARE: u8 = 0x01;
const SOFTWARE: u8 = 0x02;

pub const CLASS_KEYBOARD: u8 = 0x03;

const CH_RED: u8 = 0x01;
const CH_GREEN: u8 = 0x02;
const CH_BLUE: u8 = 0x03;

const STREAM_CHUNK: usize = 60;

const CHANNELS: [(u8, fn(&RgbColor) -> u8); 3] =
    [(CH_RED, |c| c.r), (CH_GREEN, |c| c.g), (CH_BLUE, |c| c.b)];

fn frame(payload: &[u8]) -> [u8; REPORT_SIZE] {
    let mut buf = [0u8; REPORT_SIZE];
    buf[1..1 + payload.len()].copy_from_slice(payload);
    buf
}

pub fn special_function(software: bool) -> [u8; REPORT_SIZE] {
    let mode = if software { SOFTWARE } else { HARDWARE };
    frame(&[CMD_WRITE, PROP_SPECIAL_FUNCTION, mode])
}

pub fn lighting_control(software: bool, class: u8) -> [u8; REPORT_SIZE] {
    let mode = if software { SOFTWARE } else { HARDWARE };
    frame(&[CMD_WRITE, PROP_LIGHTING_CONTROL, mode, 0x00, class])
}

pub fn firmware_query() -> [u8; REPORT_SIZE] {
    frame(&[CMD_READ, PROP_FIRMWARE])
}

fn stream_packet(nonce: u8, data: &[u8]) -> [u8; REPORT_SIZE] {
    debug_assert!(data.len() <= STREAM_CHUNK);
    let mut payload = vec![CMD_STREAM, nonce, data.len() as u8, 0x00];
    payload.extend_from_slice(data);
    frame(&payload)
}

fn submit_keyboard_24(channel: u8, packet_count: u8, finish: u8) -> [u8; REPORT_SIZE] {
    frame(&[
        CMD_WRITE,
        PROP_SUBMIT_KEYBOARD_24,
        channel,
        packet_count,
        finish,
    ])
}

pub fn color_frame_packets(colors: &[RgbColor]) -> Vec<[u8; REPORT_SIZE]> {
    let mut out = Vec::new();
    for (channel, extract) in CHANNELS {
        let plane: Vec<u8> = colors.iter().map(extract).collect();
        let chunks: Vec<&[u8]> = plane.chunks(STREAM_CHUNK).collect();
        for (i, chunk) in chunks.iter().enumerate() {
            out.push(stream_packet((i + 1) as u8, chunk));
        }
        let finish = if channel == CH_BLUE { 2 } else { 1 };
        out.push(submit_keyboard_24(channel, chunks.len() as u8, finish));
    }
    out
}

/// Key identifiers omitted from the layout-setup burst to select physical layout.
pub const SKIP_ANSI: &[u8] = &[
    0x31, 0x3f, 0x41, 0x42, 0x51, 0x53, 0x55, 0x6f, 0x7e, 0x7f, 0x80, 0x81,
];
pub const SKIP_ISO_K70_MK2: &[u8] = &[
    0x3f, 0x41, 0x42, 0x50, 0x53, 0x55, 0x6f, 0x78, 0x79, 0x7a, 0x7b, 0x7c, 0x7d, 0x7e, 0x7f, 0x80,
    0x81,
];

pub fn layout_setup_packets(skip: &[u8]) -> Vec<[u8; REPORT_SIZE]> {
    let mut out = Vec::with_capacity(LAYOUT_PACKET_COUNT + 1);
    out.push(frame(&[
        CMD_WRITE,
        PROP_LIGHTING_CONTROL,
        LAYOUT_PRIMER_MODE,
        0x00,
        0x01,
    ]));

    let mut id: u8 = 0;
    for _ in 0..LAYOUT_PACKET_COUNT {
        let mut payload = vec![CMD_WRITE, PROP_LAYOUT_SETUP, LAYOUT_SETUP_SUB, 0x00];
        for _ in 0..LAYOUT_KEYS_PER_PACKET {
            // Advance past skipped identifiers. `skip` is sorted ascending.
            for &s in skip {
                if id == s {
                    id += 1;
                }
            }
            payload.push(id);
            payload.push(LAYOUT_KEY_TRAILER);
            id += 1;
        }
        out.push(frame(&payload));
    }
    out
}

pub struct CorsairNxp<T: Transport> {
    pub(crate) transport: T,
}

impl CorsairNxp<HidTransport> {
    pub fn open(path: &str) -> Result<Self> {
        Ok(Self {
            transport: HidTransport::open(path, None, TIMEOUT_MS, false, None)?,
        })
    }
}

impl<T: Transport> CorsairNxp<T> {
    pub async fn enter_software_mode(&self, class: u8) -> Result<()> {
        let _ = self.transport.write(&firmware_query()).await;
        self.transport.write(&special_function(true)).await?;
        self.transport.write(&lighting_control(true, class)).await?;
        Ok(())
    }

    pub async fn send_layout_setup(&self, skip: &[u8]) -> Result<()> {
        let packets: Vec<Vec<u8>> = layout_setup_packets(skip)
            .into_iter()
            .map(|p| p.to_vec())
            .collect();
        self.transport.write_many(&packets).await
    }

    pub async fn leave_software_mode(&self) -> Result<()> {
        self.transport.write(&special_function(false)).await
    }

    pub async fn write_colors(&self, colors: &[RgbColor]) -> Result<()> {
        let packets: Vec<Vec<u8>> = color_frame_packets(colors)
            .into_iter()
            .map(|p| p.to_vec())
            .collect();
        self.transport.write_many(&packets).await
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use proptest::prelude::*;

    fn c(r: u8, g: u8, b: u8) -> RgbColor {
        RgbColor { r, g, b }
    }

    #[test]
    fn report_buffer_has_leading_report_id() {
        let pkt = special_function(true);
        assert_eq!(pkt.len(), REPORT_SIZE);
        assert_eq!(pkt[0], 0x00, "byte 0 is the report ID");
        assert_eq!(pkt[1], CMD_WRITE);
    }

    #[test]
    fn control_packets_match_wire_spec() {
        assert_eq!(&special_function(true)[1..4], &[0x07, 0x04, 0x02]);
        assert_eq!(&special_function(false)[1..4], &[0x07, 0x04, 0x01]);
        assert_eq!(
            &lighting_control(true, CLASS_KEYBOARD)[1..6],
            &[0x07, 0x05, 0x02, 0x00, 0x03]
        );
        assert_eq!(
            &lighting_control(false, CLASS_KEYBOARD)[1..6],
            &[0x07, 0x05, 0x01, 0x00, 0x03]
        );
        assert_eq!(&firmware_query()[1..3], &[0x0E, 0x01]);
    }

    #[test]
    fn stream_packet_carries_length_and_data() {
        let data = [0x11, 0x22, 0x33];
        let pkt = stream_packet(2, &data);
        assert_eq!(&pkt[1..5], &[0x7F, 0x02, 0x03, 0x00]);
        assert_eq!(&pkt[5..8], &data);
    }

    #[test]
    fn three_channels_committed_finish_two_on_blue() {
        // 1 key → 1 stream packet + 1 submit per channel = 6 packets.
        let pkts = color_frame_packets(&[c(10, 20, 30)]);
        assert_eq!(pkts.len(), 6);
        // Submits are the 2nd, 4th, 6th packets; finish=1 for R/G, 2 for blue.
        assert_eq!(&pkts[1][1..6], &[0x07, 0x28, CH_RED, 1, 1]);
        assert_eq!(&pkts[3][1..6], &[0x07, 0x28, CH_GREEN, 1, 1]);
        assert_eq!(&pkts[5][1..6], &[0x07, 0x28, CH_BLUE, 1, 2]);
    }

    #[test]
    fn layout_setup_has_primer_plus_four_packets() {
        let pkts = layout_setup_packets(SKIP_ISO_K70_MK2);
        assert_eq!(pkts.len(), 5);
        // Primer is a lighting-control sub-0x08 packet.
        assert_eq!(&pkts[0][1..6], &[0x07, 0x05, 0x08, 0x00, 0x01]);
        // Each of the 4 layout packets is `07 40 1E 00` then 30 `<id> C0` pairs.
        for p in &pkts[1..] {
            assert_eq!(&p[1..5], &[0x07, 0x40, 0x1E, 0x00]);
            for j in 0..LAYOUT_KEYS_PER_PACKET {
                assert_eq!(
                    p[1 + 5 + 2 * j],
                    LAYOUT_KEY_TRAILER,
                    "trailer after key {j}"
                );
            }
        }
    }

    #[test]
    fn layout_setup_skips_listed_identifiers() {
        // The running key sequence must never emit an identifier from the skip
        // list, including consecutive runs (0x79..0x7d in the ISO list).
        // Ids sit at buf[5], buf[7], …; trailers (0xC0) at buf[6], buf[8], ….
        let ids: Vec<u8> = layout_setup_packets(SKIP_ISO_K70_MK2)[1..]
            .iter()
            .flat_map(|p| (0..LAYOUT_KEYS_PER_PACKET).map(move |j| p[5 + 2 * j]))
            .collect();
        assert_eq!(ids.len(), LAYOUT_PACKET_COUNT * LAYOUT_KEYS_PER_PACKET);
        for &s in SKIP_ISO_K70_MK2 {
            assert!(
                !ids.contains(&s),
                "skip id {s:#04x} leaked into the sequence"
            );
        }
        // Emitted identifiers are strictly increasing (a monotone sequence with
        // the skipped ids removed).
        assert!(ids.windows(2).all(|w| w[0] < w[1]));
    }

    #[test]
    fn large_frame_splits_into_60_byte_stream_packets() {
        // 61 keys → 61 bytes per plane → 2 stream packets (60 + 1) + 1 submit.
        let colors = vec![c(1, 2, 3); 61];
        let pkts = color_frame_packets(&colors);
        // Red plane: packets 0,1 = stream (nonce 1,2), packet 2 = submit(count=2).
        assert_eq!(&pkts[0][1..4], &[0x7F, 1, 60]);
        assert_eq!(&pkts[1][1..4], &[0x7F, 2, 1]);
        assert_eq!(&pkts[2][1..5], &[0x07, 0x28, CH_RED, 2]);
    }

    proptest! {
        // The streamed red plane, reassembled from its stream packets, equals the
        // input red bytes in order — the encode is a faithful channel serializer.
        #[test]
        fn red_plane_round_trips(reds in prop::collection::vec(any::<u8>(), 0..200)) {
            let colors: Vec<RgbColor> = reds.iter().map(|&r| c(r, 0, 0)).collect();
            let pkts = color_frame_packets(&colors);
            // Red stream packets precede the first submit (PROP_SUBMIT at [2]).
            let mut recovered = Vec::new();
            for p in &pkts {
                if p[1] == CMD_STREAM {
                    let len = p[3] as usize;
                    recovered.extend_from_slice(&p[5..5 + len]);
                } else if p[1] == CMD_WRITE && p[2] == PROP_SUBMIT_KEYBOARD_24 {
                    break; // stop at the red submit
                }
            }
            prop_assert_eq!(recovered, reds);
        }

        // Every emitted buffer is well-formed: report ID 0, known command class.
        #[test]
        fn all_packets_well_formed(n in 0usize..300) {
            let colors = vec![c(7, 7, 7); n];
            for p in color_frame_packets(&colors) {
                prop_assert_eq!(p.len(), REPORT_SIZE);
                prop_assert_eq!(p[0], 0x00);
                prop_assert!(p[1] == CMD_STREAM || p[1] == CMD_WRITE);
            }
        }
    }
}

File: plugins/corsair_nxp.lua
-- SPDX-License-Identifier: GPL-2.0-or-later
-- SPDX-FileCopyrightText: Adam Honse (CalcProgrammer1) — OpenRGB project
--
-- Corsair NXP keyboard protocol — Lua plugin for HaloDaemon.
--
-- Port of daemon/src/drivers/vendors/corsair/protocols/corsair_nxp.rs
-- and daemon/src/drivers/vendors/corsair/devices/nxp_keyboard.rs.
--
-- Currently supports the K70 RGB MK.2 Low Profile (PID 0x1B55).
-- Adding another NXP keyboard is a new model entry in the MODELS table.

-- ── protocol constants ─────────────────────────────────────────────────────

local REPORT_SIZE = 65

local CMD_WRITE             = 0x07
local CMD_READ              = 0x0E
local CMD_STREAM            = 0x7F

local PROP_FIRMWARE         = 0x01
local PROP_SPECIAL_FUNCTION = 0x04
local PROP_LIGHTING_CONTROL = 0x05
local PROP_LAYOUT_SETUP     = 0x40
local PROP_SUBMIT_KEYBOARD  = 0x28

local LAYOUT_SETUP_SUB      = 0x1E
local LAYOUT_PRIMER_MODE    = 0x08
local LAYOUT_KEY_TRAILER    = 0xC0
local LAYOUT_KEYS_PER_PKT   = 30
local LAYOUT_PACKET_COUNT   = 4

local HARDWARE = 0x01
local SOFTWARE = 0x02

local CLASS_KEYBOARD = 0x03

local CH_RED   = 0x01
local CH_GREEN = 0x02
local CH_BLUE  = 0x03

local STREAM_CHUNK = 60

local CHANNELS = {
	{ ch = CH_RED,   extract = function(c) return c.r end },
	{ ch = CH_GREEN, extract = function(c) return c.g end },
	{ ch = CH_BLUE,  extract = function(c) return c.b end },
}

-- Key identifiers omitted from the layout-setup burst.
local SKIP_ANSI = {
	0x31, 0x3f, 0x41, 0x42, 0x51, 0x53, 0x55, 0x6f, 0x7e, 0x7f, 0x80, 0x81,
}
local SKIP_ISO_K70_MK2 = {
	0x3f, 0x41, 0x42, 0x50, 0x53, 0x55, 0x6f,
	0x78, 0x79, 0x7a, 0x7b, 0x7c, 0x7d, 0x7e, 0x7f, 0x80, 0x81,
}

local LAYOUT_SKIP = {
	["ch"] = SKIP_ISO_K70_MK2,
	["it"] = SKIP_ISO_K70_MK2,
	["us"] = SKIP_ANSI,
}

local LAYOUT_NAMES = {
	["ch"] = "Swiss (ISO)",
	["it"] = "Italian (ISO)",
	["us"] = "US (ANSI)",
}

-- ── model table ─────────────────────────────────────────────────────────────

-- K70 MK.2 LED order → device key-id. 116 entries.
local K70_MK2_KEYS = {
	0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0C, 0x0D, 0x0E, 0x0F, 0x11, 0x12,
	0x14, 0x15, 0x18, 0x19, 0x1A, 0x1B, 0x1C, 0x1D, 0x1E, 0x1F, 0x20, 0x21, 0x24, 0x25, 0x26,
	0x27, 0x28, 0x2A, 0x2B, 0x2C, 0x30, 0x31, 0x32, 0x33, 0x34, 0x35, 0x36, 0x37, 0x38, 0x39,
	0x3C, 0x3D, 0x3E, 0x3F, 0x40, 0x42, 0x43, 0x44, 0x45, 0x48, 73, 74, 75, 76, 78,
	79, 80, 81, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 96, 97,
	98, 99, 100, 101, 102, 103, 104, 105, 108, 109, 110, 111, 112, 113, 115,
	116, 117, 120, 121, 122, 123, 124, 126, 127, 128, 129, 132, 133, 134, 135,
	136, 137, 139, 140, 141, 16, 114, 47, 59, 125,
}

-- Physical key grid (7 rows × 23 columns). Cells are LED indices into
-- K70_MK2_KEYS; -1 is empty.
local NA = -1
local K70_MK2_MATRIX = {
	{ NA, NA, NA, 115, 107,  8, NA, NA, NA, NA, NA, 113, 114, NA, NA, NA, NA, NA, NA,  16, NA, NA,  NA },
	{  0, NA, 10,  18,  28, 36, NA, 46, 55, 64, 74,  NA,  84, 93, 102,  6, 15, 24, 33,  26, 35, 44,  53 },
	{  1, 11, 19,  29,  37, 47, 56, 65, 75, 85, 94,  NA, 103,  7,  25, NA, 42, 51, 60,  62, 72, 82,  91 },
	{  2, NA, 12,  20,  30, 38, NA, 48, 57, 66, 76,  86,  95, 104, 70, 80, 34, 43, 52,   9, 17, 27, 100 },
	{  3, NA, 13,  21,  31, 39, NA, 49, 58, 67, 77,  87,  96, 105, 98, 112, NA, NA, NA, 45, 54, 63,  NA },
	{  4, 111, 22, 32,  40, 50, NA, 59, NA, 68, 78,  88,  97, 106, 61, NA, NA, 81, NA,  73, 83, 92, 109 },
	{  5, 14, 23,  NA,  NA, NA, NA, 41, NA, NA, NA,  NA,  69, 79,  89, 71, 90, 99, 108, 101, NA, 110, NA },
}

local MODELS = {
	[0x1B55] = {
		name = "K70 RGB MK.2 Low Profile",
		model = "K70 RGB MK.2 Low Profile",
		keys = K70_MK2_KEYS,
		matrix = K70_MK2_MATRIX,
		wire_len = 144,
	},
}

-- ── protocol helpers ───────────────────────────────────────────────────────

local function frame(payload)
	local buf = halod.buffer(REPORT_SIZE)
	for i, b in ipairs(payload) do
		buf:set_u8(i, b)  -- byte 0 stays 0x00 (report id); payload at [1..]
	end
	return buf
end

local function special_function(software)
	local mode = software and SOFTWARE or HARDWARE
	return frame({ CMD_WRITE, PROP_SPECIAL_FUNCTION, mode })
end

local function lighting_control(software, class)
	local mode = software and SOFTWARE or HARDWARE
	return frame({ CMD_WRITE, PROP_LIGHTING_CONTROL, mode, 0x00, class })
end

local function firmware_query()
	return frame({ CMD_READ, PROP_FIRMWARE })
end

local function stream_packet(nonce, data)
	local payload = { CMD_STREAM, nonce, #data, 0x00 }
	for _, b in ipairs(data) do
		payload[#payload + 1] = b
	end
	return frame(payload)
end

local function submit_keyboard_24(channel, packet_count, finish)
	return frame({ CMD_WRITE, PROP_SUBMIT_KEYBOARD, channel, packet_count, finish })
end

-- Encode a wire buffer (indexed by device key-id) into the NXP stream packets.
local function color_frame_packets(colors)
	local out = {}
	for _, chdef in ipairs(CHANNELS) do
		local plane = {}
		for _, c in ipairs(colors) do
			plane[#plane + 1] = chdef.extract(c)
		end
		-- chunk into STREAM_CHUNK (60) groups
		local chunks = {}
		for i = 1, #plane, STREAM_CHUNK do
			local chunk = {}
			for j = i, math.min(i + STREAM_CHUNK - 1, #plane) do
				chunk[#chunk + 1] = plane[j]
			end
			chunks[#chunks + 1] = chunk
		end
		for i, chunk in ipairs(chunks) do
			out[#out + 1] = stream_packet(i, chunk)
		end
		local finish = (chdef.ch == CH_BLUE) and 2 or 1
		out[#out + 1] = submit_keyboard_24(chdef.ch, #chunks, finish)
	end
	return out
end

-- Build the layout-setup burst (5 packets: primer + 4×30 key-ids).
local function layout_setup_packets(skip)
	local out = {}
	out[#out + 1] = frame({ CMD_WRITE, PROP_LIGHTING_CONTROL, LAYOUT_PRIMER_MODE, 0x00, 0x01 })

	-- Build skip set for O(1) lookup.
	local skip_set = {}
	for _, s in ipairs(skip) do
		skip_set[s] = true
	end

	local id = 0
	for _ = 1, LAYOUT_PACKET_COUNT do
		local payload = { CMD_WRITE, PROP_LAYOUT_SETUP, LAYOUT_SETUP_SUB, 0x00 }
		for _ = 1, LAYOUT_KEYS_PER_PKT do
			while skip_set[id] do
				id = id + 1
			end
			payload[#payload + 1] = id
			payload[#payload + 1] = LAYOUT_KEY_TRAILER
			id = id + 1
		end
		out[#out + 1] = frame(payload)
	end
	return out
end

-- ── model helpers ───────────────────────────────────────────────────────────

local function model_for(pid)
	return MODELS[pid]
end

-- Compute LED positions from the physical matrix.
local function led_positions(spec)
	local rows = #spec.matrix
	local cols = #spec.matrix[1]
	local nkeys = #spec.keys
	local positions = {}
	for i = 1, nkeys do
		positions[i] = { id = i - 1, x = 0.5, y = 0.5 }
	end
	for r = 1, rows do
		for c = 1, cols do
			local cell = spec.matrix[r][c]
			if cell >= 0 then
				local idx = cell + 1 -- Lua is 1-indexed
				if positions[idx] then
					positions[idx].x = (c - 1) / (cols - 1)
					positions[idx].y = (r - 1) / (rows - 1)
				end
			end
		end
	end
	return positions
end

-- ── per-device helpers (called with dev and its spec) ──────────────────────

-- Scatter LED-ordered colors into the wire buffer indexed by device key-id.
local function wire_buffer(spec, led_colors)
	local buf = {}
	for _ = 1, spec.wire_len do
		buf[#buf + 1] = { r = 0, g = 0, b = 0 }
	end
	for i, key_id in ipairs(spec.keys) do
		local color = led_colors[i]
		if color then
			buf[key_id + 1] = color -- key_id → 1-based index
		end
	end
	return buf
end

-- ── plugin table ────────────────────────────────────────────────────────────

-- The plugin stores per-device state keyed by the device's transport path.
-- In the plugin model each callback receives `dev`; we stash the model spec
-- and current layout in the dev table so they survive across calls.
-- Actually, `dev` is re-created per call, so we use a weak-keyed cache.
local device_state = setmetatable({}, { __mode = "k" })

local function ensure_state(dev)
	if not device_state[dev] then
		local spec = model_for(dev.match.pid)
		device_state[dev] = {
			spec = spec,
			layout = "ch",  -- default Swiss ISO
		}
	end
	return device_state[dev]
end

return {
	match = {
		transport = "hid",
		vid = 0x1B1C,
		pid = 0x1B55,
		interface = 1,
	},

	identity = {
		vendor = "Corsair",
		model = "K70 RGB MK.2 Low Profile",
		name = "Corsair K70 RGB MK.2 Low Profile",
		author = "HaloDaemon (ported from OpenRGB)",
		version = "1.0.0",
		description = "Corsair NXP keyboard protocol driver. Supports the K70 RGB MK.2 Low Profile.",
	},

	transports = { hid = { report_size = 0, timeout_ms = 1000 } },

	rgb = {
		zones = {
			{
				id = "keyboard",
				name = "Keyboard",
				topology = { type = "grid" },
				leds = led_positions(MODELS[0x1B55]),
			},
		},
	},

	choice = {
		choices = {
			{
				key = "layout",
				label = "Physical Layout",
				category = "Keyboard",
				display = "list",
				options = {
					{ id = "ch", label = "Swiss (ISO)" },
					{ id = "it", label = "Italian (ISO)" },
					{ id = "us", label = "US (ANSI)" },
				},
				default = 0,
			},
		},
	},

	-- ── callbacks ─────────────────────────────────────────────────────────

	initialize = function(dev)
		local st = ensure_state(dev)
		local spec = st.spec
		if not spec then
			log("corsair_nxp: unknown pid " .. string.format("0x%04x", dev.match.pid))
			return false
		end

		-- Enter software/direct control mode.
		dev.transport:write(firmware_query())
		dev.transport:write(special_function(true))
		dev.transport:write(lighting_control(true, CLASS_KEYBOARD))

		-- Send layout setup.
		local skip = LAYOUT_SKIP[st.layout] or SKIP_ISO_K70_MK2
		local pkts = layout_setup_packets(skip)
		dev.transport:write_many(pkts)

		log("corsair_nxp: " .. spec.name .. " initialized")
		return true
	end,

	close = function(dev)
		dev.transport:write(special_function(false))
	end,

	write_frame = function(dev, zone_id, colors)
		if zone_id ~= "keyboard" then return end
		local st = ensure_state(dev)
		local wbuf = wire_buffer(st.spec, colors)
		local pkts = color_frame_packets(wbuf)
		dev.transport:write_many(pkts)
	end,

	apply = function(dev, state)
		local st = ensure_state(dev)
		local spec = st.spec
		local led_count = #spec.keys

		if state.mode == "static" then
			local c = state.color
			local colors = {}
			for _ = 1, led_count do
				colors[#colors + 1] = { r = c.r, g = c.g, b = c.b }
			end
			local wbuf = wire_buffer(spec, colors)
			dev.transport:write_many(color_frame_packets(wbuf))
		elseif state.mode == "per_led" then
			-- Per-LED map keyed by LED index string.
			local colors = {}
			local black = { r = 0, g = 0, b = 0 }
			for i = 0, led_count - 1 do
				local key = tostring(i)
				local c = (state.zones and state.zones["keyboard"] and state.zones["keyboard"][key]) or black
				colors[#colors + 1] = { r = c.r, g = c.g, b = c.b }
			end
			local wbuf = wire_buffer(spec, colors)
			dev.transport:write_many(color_frame_packets(wbuf))
		end
		-- "engine" mode is handled by write_frame; "native_effect" / "direct_effect" are no-ops.
	end,

	set_choice = function(dev, key, selected)
		if key ~= "layout" then return end
		local st = ensure_state(dev)
		local layouts = { "ch", "it", "us" }
		local new_layout = layouts[selected + 1] or "ch"
		st.layout = new_layout

		local skip = LAYOUT_SKIP[new_layout] or SKIP_ISO_K70_MK2
		dev.transport:write_many(layout_setup_packets(skip))
		log("corsair_nxp: layout set to " .. (LAYOUT_NAMES[new_layout] or new_layout))
	end,
}
